src/NuGet.Services.AzureSearch/Auxiliary2AzureSearch/Auxiliary2AzureSearchCommand.cs

```diff
@@ -1,17 +1,287 @@
+                {
+                    documentsToPush = await PushIndexActionsAsync(indexActionsToPush, documentsToPush, timeSinceLastPush);
+                }
+
+                // If all of the IDs to process plus the current ID were to need a document for each search filter and
+                // that number plus the current known documents to push would make the batch larger than the maximum
+                // batch size, produce index actions for the IDs that we have collected so far.
+                if (((idsToProcess.Count + 1) * SearchFilterCount) + documentsToPush > _options.Value.AzureSearchBatchSize)
+                {
+                    documentsToPush = await ProcessIdsAsync(idsToProcess, indexActionsToPush, documentsToPush, changes);
+                }
+
+                idsToProcess.Add(idToBatch);
+            }
+
+            // Process any leftover IDs that didn't make it into a full batch.
+            if (idsToProcess.Count > 0)
+            {
+                documentsToPush = await ProcessIdsAsync(idsToProcess, indexActionsToPush, documentsToPush, changes);
+            }
+
+            // Push any leftover index actions that didn't make it into a full batch.
+            if (documentsToPush > 0)
+            {
+                documentsToPush = await PushIndexActionsAsync(indexActionsToPush, documentsToPush, timeSinceLastPush);
+            }
+
+            Guard.Assert(idsToProcess.IsEmpty, "There should be no more IDs to process.");
+            Guard.Assert(indexActionsToPush.IsEmpty, "There should be no more index actions to push.");
+            Guard.Assert(documentsToPush == 0, "There should be no more documents to push.");
+        }
+
+        private async Task<int> ProcessIdsAsync(
+            ConcurrentBag<string> idsToProcess,
+            ConcurrentBag<IdAndValue<IndexActions>> indexActionsToPush,
+            int documentsToPush,
+            SortedDictionary<string, long> changes)
+        {
+            await ParallelAsync.Repeat(
+                async () =>
+                {
+                    while (idsToProcess.TryTake(out var id))
+                    {
+                        var indexActions = await _indexActionBuilder.UpdateAsync(
+                            id,
+                            sf => _searchDocumentBuilder.UpdateDownloadCount(id, sf, changes[id]));
+
+                        if (indexActions.IsEmpty)
+                        {
+                            continue;
+                        }
+
+                        Guard.Assert(indexActions.Hijack.Count == 0, "There should be no hijack index changes.");
+
+                        indexActionsToPush.Add(new IdAndValue<IndexActions>(id, indexActions));
+                        Interlocked.Add(ref documentsToPush, indexActions.Search.Count);
+                    }
+                },
+                _options.Value.MaxConcurrentVersionListWriters);
```

Comment:
`MaxConcurrentVersionListWriters` makes it sound like only `MaxConcurrentVersionListWriters` tasks may access the version concurrently, but it's actually `MaxConcurrentVersionListWriters *  MaxConcurrentBatches`. What do you think of renaming this variable to something like `MaxConcurrentVersionListWritersPerBatch`? (I know this is an annoying change since you have to update NuGetDeployment too, sorry :( )

Reply:
Well it varies actually. In Catalog2AzureSearch there is only one thread using `IBatchPusher` so `MaxConcurrentVersionListWriters` is correctly named. `MaxConcurrentBatches` is used in that case to process package IDs in parallel and yield index changes to be pushed later by a single guy.

In the other jobs you're right it's more `MaxConcurrentVersionListWritersPerBatch`. Early on I thought about coming up with separate parallelism controls per job so that they would be named more clearly but this became not worth IMO the effort since the `BatchPusher` implementation was used everywhere and needed a single definition of `MaxConcurrentVersionListWriters` so the code could be shared.

In short I feel that modification of parallelism values should be done with full cognizance of the effects on the particular job and the name is only the very first step. Hopefully it's "close enough" to get rid of most of the confusion.

Perhaps a code comment on this property in `AzureSearchJobConfiguration` with this sort of detail would be enough?

Reply:
Oh another thing to add is that in practice I think Auxiliary2AzureSearch will have `MaxConcurrentBatches` set to 1. We don't want this job to cause too much load on the search index.

Reply:
> Perhaps a code comment on this property in `AzureSearchJobConfiguration` with this sort of detail would be enough?

That works for me